Style and performance tweaks to CSS property name and keyword value tables

darinadler opened the pull request:
#### a10c81dcc8fb22521e0999dc7e119a94c92c7520
<pre>
Style and performance tweaks to CSS property name and keyword value tables
<a href="https://bugs.webkit.org/show_bug.cgi?id=247244">https://bugs.webkit.org/show_bug.cgi?id=247244</a>
rdar://101730654

Reviewed by Sam Weinig.

* Source/WebCore/animation/CSSPropertyAnimation.cpp:
(WebCore::CSSPropertyAnimationWrapperMap::CSSPropertyAnimationWrapperMap):
Use nameLiteral, numCSSPRoperties instead of lastCSSProperty, and
allCSSProperties.

* Source/WebCore/animation/CSSTransition.cpp:
(WebCore::CSSTransition::createEvent): Use nameString.

* Source/WebCore/animation/CSSTransition.h: Use const AtomString&amp; for the
return value of transitionProperty to cut down on reference count churn.
Use nameString, moved isCSSTransition to private.

* Source/WebCore/animation/KeyframeEffect.cpp:
(WebCore::KeyframeEffect::CSSPropertyIDToIDLAttributeName): Use nameForIDL.
(WebCore::processKeyframeLikeObject): Use isExposed.

* Source/WebCore/css/CSSBackgroundRepeatValue.cpp:
(WebCore::CSSBackgroundRepeatValue::customCSSText const): Use nameString.
* Source/WebCore/css/CSSComputedStyleDeclaration.cpp:
(WebCore::CSSComputedStyleDeclaration::item const): Ditto.
* Source/WebCore/css/CSSCustomPropertyValue.cpp:
(WebCore::CSSCustomPropertyValue::customCSSText const): Ditto.

* Source/WebCore/css/CSSFunctionValue.cpp:
(WebCore::CSSFunctionValue::customCSSText const): Use nameLiteral.
* Source/WebCore/css/CSSGridAutoRepeatValue.cpp:
(WebCore::CSSGridAutoRepeatValue::customCSSText const): Ditto.

* Source/WebCore/css/CSSPrimitiveValue.cpp:
(WebCore::propertyName): Deleted.
(WebCore::valueName): Deleted.
(WebCore::CSSPrimitiveValue::stringValue const): Use nameString.
(WebCore::CSSPrimitiveValue::formatNumberForCustomCSSText const): Ditto.
(WebCore::CSSPrimitiveValue::equals const): Compare property and value
IDs rather than converting to strings and comparing the strings.

* Source/WebCore/css/CSSStyleDeclaration.cpp:
(WebCore::parseJavaScriptCSSPropertyName): Use AtomString for the
map keys for better performance. Don&apos;t add a trailing NUL character,
since it&apos;s not required any more. Use findCSSProperty.
(WebCore::lookupCSSPropertyFromIDLAttribute): Don&apos;t add a trailing
NUL character since it&apos;s not required any more. Use findCSSProperty.
(WebCore::CSSStyleDeclaration::getCSSPropertyIDFromJavaScriptPropertyName):
Merged propertyIDFromJavaScriptCSSPropertyName in here, we don&apos;t need
two separate functions. Use isExposed.

* Source/WebCore/css/CSSValuePool.cpp:
(WebCore::StaticCSSValuePool::StaticCSSValuePool): Use allCSSValueKeywords.
(WebCore::CSSValuePool::createIdentifierValue): Do the assertion without
using firstCSSValueKeyword and lastCSSValueKeyword.

* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::createTransitionPropertyValue): Use nameString.
(WebCore::ComputedStyleExtractor::propertyValue): Use isExposed.
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle): Ditto.
(WebCore::ComputedStyleExtractor::copyProperties): Use allLonghandCSSProperties.

* Source/WebCore/css/DOMCSSNamespace.cpp:
(WebCore::DOMCSSNamespace::supports): Use isExposed.

* Source/WebCore/css/PropertySetCSSStyleDeclaration.cpp:
(WebCore::PropertySetCSSStyleDeclaration::length const): Use isExposed.
(WebCore::PropertySetCSSStyleDeclaration::item const): Ditto.
(WebCore::PropertySetCSSStyleDeclaration::getPropertyCSSValue): Ditto.
(WebCore::PropertySetCSSStyleDeclaration::getPropertyValue): Ditto.
(WebCore::PropertySetCSSStyleDeclaration::getPropertyPriority): Ditto.
(WebCore::PropertySetCSSStyleDeclaration::getPropertyShorthand): Ditto.
(WebCore::PropertySetCSSStyleDeclaration::isPropertyImplicit): Ditto.
(WebCore::PropertySetCSSStyleDeclaration::setProperty): Ditto.
(WebCore::PropertySetCSSStyleDeclaration::removeProperty): Ditto.
(WebCore::PropertySetCSSStyleDeclaration::getPropertyValueInternal): Ditto.
(WebCore::PropertySetCSSStyleDeclaration::setPropertyInternal): Ditto.
(WebCore::PropertySetCSSStyleDeclaration::isExposed const): Renamed from
isCSSPropertyExposed and added the check for CSSPropertyInvalid here so
we don&apos;t need to repeat it at all callers.
* Source/WebCore/css/PropertySetCSSStyleDeclaration.h: Updated for above.

* Source/WebCore/css/StyleColor.cpp:
(WebCore::StyleColor::colorFromAbsoluteKeyword): Use nameLiteral.

* Source/WebCore/css/StyleProperties.cpp:
(WebCore::StyleProperties::fontValue const): Use nameString and nameLiteral.
(WebCore::StyleProperties::get2Values const): Use nameString.
(WebCore::StyleProperties::get4Values const): Use nameString.
(WebCore::StyleProperties::getLayeredShorthandValue const): Use nameLiteral.
(WebCore::StyleProperties::getPropertyShorthand const): Use nameString.
(WebCore::StyleProperties::asTextInternal const): Use isLonghand and nameLiteral.
(WebCore::MutableStyleProperties::setProperty): Use isExposed and isInternal.
(WebCore::StyleProperties::PropertyReference::cssName const): Use nameString.

* Source/WebCore/css/makeprop.pl: Removed support for runtime-flag. We don&apos;t need it any more.
Moved the computedPropertyIDs array out of the header. Moved the hash table structure in here
from HashTools.h. Set the %7bit flag since all the clients already forbid non-ASCII characters.
Removed unneeded custom lookup-function-name, hash-function-name, and word-array-name since
class-name already prevents collisions. Added isLonghand. Renamed isInternalCSSProperty to just
isInternal since the argument is a strongly typed CSSPRopertyID. Similarly renamed
isCSSPropertyExposed to isExposed. Replaced the findProperty function with one named
findCSSProperty that returns a CSSPropertyID insted of a pointer into the hash table. Renamed
getPropertyName to nameLiteral. Renamed getPropertyNameAtomString to nameString. Use a
NeverDestroyed instead of allocating the array of AtomString on the heap. Got rid of
getPropertyNameString. Renamed getJSPropertyName to nameForIDL. Made some tables constexpr.
Renamed getRelatedPropertyID to relatedProperty. Made the property index constants use uint16_t
and the string length constant use unsigned instead of size_t. Use constexpr for them. Changed
computedPropertyIDs to be a std::array. Removed convertToCSSPropertyID. Added CSSPropertiesRange
class template to help us implement range-based for statements. Added allCSSProperties and
allLonghandCSSProperties.

* Source/WebCore/css/makevalues.pl: Moved the hash table structure in here
from HashTools.h. Set the %7bit flag since all the clients already forbid
non-ASCII characters. Removed unneeded custom lookup-function-name,
hash-function-name, and word-array-name since class-name already prevents collisions.
Replaced the findValue function with one named findCSSValueKeyword that returns
a CSSValueID instead of a pointer into the hash table. Renamed getValueName
to nameLiteral. Renamed getValueNameAtomString to nameString. Got rid of
getValueNameString. Made the index constants use uint16_t and the string length
constant use unsigned instead of size_t. Added AllCSSValueKeywordsRange class template
to help us implement range-based for statements. Added allCSSValueKeywords.

* Source/WebCore/css/parser/CSSParserFastPaths.cpp:
(WebCore::CSSParserFastPaths::maybeParseValue): Use isExposed.

* Source/WebCore/css/parser/CSSParserImpl.cpp:
(WebCore::CSSParserImpl::consumeDeclaration): Use isExposed.

* Source/WebCore/css/parser/CSSPropertyParser.cpp: Removed include of
HashTools.h.
(WebCore::cssPropertyID): Removed unneeded extra characters in buffer. Removed
code that adds a trailing NUL character. Use findCSSProperty.
(WebCore::isAppleLegacyCSSValueKeyword): Removed unnecessarily complex handling
of &quot;-apple-wireless&quot; prefix, simplifying making it match the other cases. Also
tweaked the logic to do less excess checking.
(WebCore::cssValueKeywordID): Removed code that adds a trailing NUL character.
SImplified the isAppleLegacyCSSValueKeyword section a little and removed a
misplaced comment. Use findCSSValueKeyword.
(WebCore::CSSPropertyParser::addProperty): Use isExposed.
(WebCore::consumeWillChange): Use isExposed.
(WebCore::CSSPropertyParser::parseSingleValue): Use isExposed.
(WebCore::CSSPropertyParser::parseCounterStyleDescriptor): Use isExposed.
(WebCore::CSSPropertyParser::parseFontFaceDescriptor): Use isExposed.
(WebCore::CSSPropertyParser::parseFontPaletteValuesDescriptor): Use isExposed.
(WebCore::CSSPropertyParser::consumeContainIntrinsicSizeShorthand): Use isExposed.

* Source/WebCore/css/typedom/ComputedStylePropertyMapReadOnly.cpp:
(WebCore::ComputedStylePropertyMapReadOnly::get const): Use isExposed.
(WebCore::ComputedStylePropertyMapReadOnly::getAll const): Use isExposed.
(WebCore::ComputedStylePropertyMapReadOnly::entries const): Use nameString.

* Source/WebCore/css/typedom/MainThreadStylePropertyMapReadOnly.cpp:
(WebCore::MainThreadStylePropertyMapReadOnly::get const): Use isExposed.
(WebCore::MainThreadStylePropertyMapReadOnly::getAll const): Use isExposed.

* Source/WebCore/css/typedom/StylePropertyMap.cpp:
(WebCore::StylePropertyMap::remove): Use isExposed.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::exposedComputedCSSPropertyIDs): Tweaked code since
computedPropertyIDs is now a std::array. Also decreased copying a little
using std::optional::emplace. We could do even better if we made a FixedVector
constructor that took a pair of iterators or a Span. Use isExposed.
(WebCore::Document::detachFromFrame): Fixed comment typo.

* Source/WebCore/html/CustomPaintImage.cpp: Use nameString.

* Source/WebCore/inspector/InspectorStyleSheet.cpp:
(WebCore::InspectorStyle::collectProperties const): Use allCSSProperties and isExposed.
Fixes a bug where we would not iterate over the last CSS property. Use nameString.
(WebCore::InspectorStyle::styleWithProperties const): Use isExposed and nameString.

* Source/WebCore/inspector/agents/InspectorAnimationAgent.cpp:
(WebCore::buildObjectForKeyframes): Use nameString.

* Source/WebCore/inspector/agents/InspectorCSSAgent.cpp:
(WebCore::InspectorCSSAgent::getSupportedCSSProperties): Use allCSSProperties,
isExposed, nameString, and allCSSValueKeywords.

* Source/WebCore/page/CaptionUserPreferencesMediaAF.cpp:
(WebCore::CaptionUserPreferencesMediaAF::captionsWindowCSS const): Use nameLiteral.
(WebCore::appendCSS): Turned this into a template to cut down on temporary strings.
(WebCore::CaptionUserPreferencesMediaAF::windowRoundedCornerRadiusCSS const):
Reorder arguments and stop using nameString when calling appendCSS.
(WebCore::CaptionUserPreferencesMediaAF::colorPropertyCSS const): Ditto.
(WebCore::CaptionUserPreferencesMediaAF::captionsTextEdgeCSS const): Ditto.
Also got rid of NeverDestroyed&lt;const String&gt; because we are appending these strings,
so we can just use literals.

* Source/WebCore/platform/HashTools.h: Remove Property, Value, findProperty, and findValue.

* Source/WebCore/platform/animation/Animation.cpp:
(WebCore::operator&lt;&lt;): Use nameLiteral.
* Source/WebCore/rendering/style/RenderStyleConstants.cpp:
(WebCore::operator&lt;&lt;): Ditto.

* Source/WebCore/style/PropertyCascade.cpp:
(WebCore::Style::PropertyCascade::lastDeferredPropertyResolvingRelated const):
Use relatedProperty.

* Source/WebCore/style/StyleBuilderCustom.h:
(WebCore::Style::BuilderCustom::applyValueWillChange): Use isExposed.

Canonical link: <a href="https://commits.webkit.org/256223@main">https://commits.webkit.org/256223@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/07b0a8745a575f6f44328ab161837a056ecdb09d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/95083 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/4226 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/28000 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/104684 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/164935 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/99073 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/4317 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/32412 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/87394 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/100584 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/100750 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/3140 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/81614 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/30128 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/85052 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/86886 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/73024 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/38816 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/18441 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/36638 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/19722 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4304 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/40572 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/42417 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/42551 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/38963 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

weinig commented on Source/WebCore/css/ComputedStyleExtractor.cpp:
```diff
@@ -4350,9 +4350,8 @@ Ref<MutableStyleProperties> ComputedStyleExtractor::copyPropertiesInSet(const CS
-    for (unsigned i = firstCSSProperty; i < firstShorthandProperty; ++i) {
-        auto propertyID = convertToCSSPropertyID(i);
+    list.reserveInitialCapacity(numCSSPropertyLonghands);
+    for (auto propertyID : allLonghandCSSProperties()) {
```
Not necessarily something to actually change, but what is your feeling on this:

```c++
auto list = allLonghandCSSProperties().compactMap([](auto propertyID) -> std::optional<CSSProperty> {
    if (auto value = propertyValue(propertyID))
        return CSSProperty(propertyID, WTFMove(value)));
    return std::nullopt;
});
```

(though, I think it would probably be `auto list = WTF::compactMap(allLonghandCSSProperties(), ...` since Vector doesn't have a `compactMap` of its own).

darinadler replied:
I love that; I am tempted to follow up with it.

weinig commented on Source/WebCore/css/PropertySetCSSStyleDeclaration.cpp:
```diff
@@ -153,7 +153,7 @@ unsigned PropertySetCSSStyleDeclaration::length() const
     unsigned exposed = 0;
     for (unsigned i = 0; i < m_propertySet->propertyCount(); i++) {
-        if (isCSSPropertyExposed(m_propertySet->propertyAt(i).id()))
+        if (isExposed(m_propertySet->propertyAt(i).id()))
```
`StyleProperties` would be another good one to add an iterator for at some point.

darinadler replied:
Great idea.

weinig commented on Source/WebCore/css/makeprop.pl:
```diff
@@ -427,18 +424,17 @@ sub sortByDescendingPriorityAndName
 #include \"CSSPropertyNames.h\"
-#include \"HashTools.h\"
+
+#include \"CSSProperty.h\"
```
Any idea why only some of these quote #includes are escaped?

darinadler replied:
None whatsoever. I didn’t even notice!

weinig commented on Source/WebCore/css/makeprop.pl:
```diff
@@ -427,18 +424,17 @@ sub sortByDescendingPriorityAndName
-#include \"HashTools.h\"
+
+#include \"CSSProperty.h\"
 #include "DeprecatedGlobalSettings.h"
```
I bet you can remove this now that we don't have $runtimeFlags.

weinig commented on Source/WebCore/css/makeprop.pl:
```diff
@@ -1074,30 +1035,48 @@ sub sortWithPrefixedPropertiesLast
+        constexpr Iterator& operator++() { ++index; return *this; }
+        constexpr bool operator!=(std::nullptr_t) const { return index <= static_cast<uint16_t>(last); }
+    };
+    static constexpr Iterator begin() { return { }; }
```
Might be a good idea to add a `size()` function that returns `last - first` so that the sized container algorithms can be a bit more efficient.

darinadler replied:
I like that idea. I just learned about the `std::ranges` concept `sized_range`.

But is there somewhere I would use that in this patch?

darinadler replied:
It’s `last - first + 1`, because `last + 1 == end`.

weinig replied:
I had a sinking suspicion I would have an off by one there :).

weinig commented on Source/WebCore/css/makevalues.pl:
```diff
@@ -193,31 +187,40 @@
+        constexpr Iterator& operator++() { ++index; return *this; }
+        constexpr bool operator!=(std::nullptr_t) const { return index < numCSSValueKeywords; }
+    };
+    static constexpr Iterator begin() { return { }; }
```
Same thought about a `size()` function, but this one could just return `numCSSValueKeywords`

darinadler committed:
Style and performance tweaks to CSS property name and keyword value tables
https://bugs.webkit.org/show_bug.cgi?id=247244
rdar://101730654

Reviewed by Sam Weinig.

* Source/WebCore/animation/CSSPropertyAnimation.cpp:
(WebCore::CSSPropertyAnimationWrapperMap::CSSPropertyAnimationWrapperMap):
Use nameLiteral, numCSSPRoperties instead of lastCSSProperty, and
allCSSProperties.

* Source/WebCore/animation/CSSTransition.cpp:
(WebCore::CSSTransition::createEvent): Use nameString.

* Source/WebCore/animation/CSSTransition.h: Use const AtomString& for the
return value of transitionProperty to cut down on reference count churn.
Use nameString, moved isCSSTransition to private.

* Source/WebCore/animation/KeyframeEffect.cpp:
(WebCore::KeyframeEffect::CSSPropertyIDToIDLAttributeName): Use nameForIDL.
(WebCore::processKeyframeLikeObject): Use isExposed.

* Source/WebCore/css/CSSBackgroundRepeatValue.cpp:
(WebCore::CSSBackgroundRepeatValue::customCSSText const): Use nameString.
* Source/WebCore/css/CSSComputedStyleDeclaration.cpp:
(WebCore::CSSComputedStyleDeclaration::item const): Ditto.
* Source/WebCore/css/CSSCustomPropertyValue.cpp:
(WebCore::CSSCustomPropertyValue::customCSSText const): Ditto.

* Source/WebCore/css/CSSFunctionValue.cpp:
(WebCore::CSSFunctionValue::customCSSText const): Use nameLiteral.
* Source/WebCore/css/CSSGridAutoRepeatValue.cpp:
(WebCore::CSSGridAutoRepeatValue::customCSSText const): Ditto.

* Source/WebCore/css/CSSPrimitiveValue.cpp:
(WebCore::propertyName): Deleted.
(WebCore::valueName): Deleted.
(WebCore::CSSPrimitiveValue::stringValue const): Use nameString.
(WebCore::CSSPrimitiveValue::formatNumberForCustomCSSText const): Ditto.
(WebCore::CSSPrimitiveValue::equals const): Compare property and value
IDs rather than converting to strings and comparing the strings.

* Source/WebCore/css/CSSStyleDeclaration.cpp:
(WebCore::parseJavaScriptCSSPropertyName): Use AtomString for the
map keys for better performance. Don't add a trailing NUL character,
since it's not required any more. Use findCSSProperty.
(WebCore::lookupCSSPropertyFromIDLAttribute): Don't add a trailing
NUL character since it's not required any more. Use findCSSProperty.
(WebCore::CSSStyleDeclaration::getCSSPropertyIDFromJavaScriptPropertyName):
Merged propertyIDFromJavaScriptCSSPropertyName in here, we don't need
two separate functions. Use isExposed.

* Source/WebCore/css/CSSValuePool.cpp:
(WebCore::StaticCSSValuePool::StaticCSSValuePool): Use allCSSValueKeywords.
(WebCore::CSSValuePool::createIdentifierValue): Do the assertion without
using firstCSSValueKeyword and lastCSSValueKeyword.

* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::createTransitionPropertyValue): Use nameString.
(WebCore::ComputedStyleExtractor::propertyValue): Use isExposed.
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle): Ditto.
(WebCore::ComputedStyleExtractor::copyProperties): Use allLonghandCSSProperties.

* Source/WebCore/css/DOMCSSNamespace.cpp:
(WebCore::DOMCSSNamespace::supports): Use isExposed.

* Source/WebCore/css/PropertySetCSSStyleDeclaration.cpp:
(WebCore::PropertySetCSSStyleDeclaration::length const): Use isExposed.
(WebCore::PropertySetCSSStyleDeclaration::item const): Ditto.
(WebCore::PropertySetCSSStyleDeclaration::getPropertyCSSValue): Ditto.
(WebCore::PropertySetCSSStyleDeclaration::getPropertyValue): Ditto.
(WebCore::PropertySetCSSStyleDeclaration::getPropertyPriority): Ditto.
(WebCore::PropertySetCSSStyleDeclaration::getPropertyShorthand): Ditto.
(WebCore::PropertySetCSSStyleDeclaration::isPropertyImplicit): Ditto.
(WebCore::PropertySetCSSStyleDeclaration::setProperty): Ditto.
(WebCore::PropertySetCSSStyleDeclaration::removeProperty): Ditto.
(WebCore::PropertySetCSSStyleDeclaration::getPropertyValueInternal): Ditto.
(WebCore::PropertySetCSSStyleDeclaration::setPropertyInternal): Ditto.
(WebCore::PropertySetCSSStyleDeclaration::isExposed const): Renamed from
isCSSPropertyExposed and added the check for CSSPropertyInvalid here so
we don't need to repeat it at all callers.
* Source/WebCore/css/PropertySetCSSStyleDeclaration.h: Updated for above.

* Source/WebCore/css/StyleColor.cpp:
(WebCore::StyleColor::colorFromAbsoluteKeyword): Use nameLiteral.

* Source/WebCore/css/StyleProperties.cpp:
(WebCore::StyleProperties::fontValue const): Use nameString and nameLiteral.
(WebCore::StyleProperties::get2Values const): Use nameString.
(WebCore::StyleProperties::get4Values const): Use nameString.
(WebCore::StyleProperties::getLayeredShorthandValue const): Use nameLiteral.
(WebCore::StyleProperties::getPropertyShorthand const): Use nameString.
(WebCore::StyleProperties::asTextInternal const): Use isLonghand and nameLiteral.
(WebCore::MutableStyleProperties::setProperty): Use isExposed and isInternal.
(WebCore::StyleProperties::PropertyReference::cssName const): Use nameString.

* Source/WebCore/css/makeprop.pl: Removed support for runtime-flag. We don't need it any more.
Moved the computedPropertyIDs array out of the header. Moved the hash table structure in here
from HashTools.h. Set the %7bit flag since all the clients already forbid non-ASCII characters.
Removed unneeded custom lookup-function-name, hash-function-name, and word-array-name since
class-name already prevents collisions. Added isLonghand. Renamed isInternalCSSProperty to just
isInternal since the argument is a strongly typed CSSPRopertyID. Similarly renamed
isCSSPropertyExposed to isExposed. Replaced the findProperty function with one named
findCSSProperty that returns a CSSPropertyID insted of a pointer into the hash table. Renamed
getPropertyName to nameLiteral. Renamed getPropertyNameAtomString to nameString. Use a
NeverDestroyed instead of allocating the array of AtomString on the heap. Got rid of
getPropertyNameString. Renamed getJSPropertyName to nameForIDL. Made some tables constexpr.
Renamed getRelatedPropertyID to relatedProperty. Made the property index constants use uint16_t
and the string length constant use unsigned instead of size_t. Use constexpr for them. Changed
computedPropertyIDs to be a std::array. Removed convertToCSSPropertyID. Added CSSPropertiesRange
class template to help us implement range-based for statements. Added allCSSProperties and
allLonghandCSSProperties.

* Source/WebCore/css/makevalues.pl: Moved the hash table structure in here
from HashTools.h. Set the %7bit flag since all the clients already forbid
non-ASCII characters. Removed unneeded custom lookup-function-name,
hash-function-name, and word-array-name since class-name already prevents collisions.
Replaced the findValue function with one named findCSSValueKeyword that returns
a CSSValueID instead of a pointer into the hash table. Renamed getValueName
to nameLiteral. Renamed getValueNameAtomString to nameString. Got rid of
getValueNameString. Made the index constants use uint16_t and the string length
constant use unsigned instead of size_t. Added AllCSSValueKeywordsRange class template
to help us implement range-based for statements. Added allCSSValueKeywords.

* Source/WebCore/css/parser/CSSParserFastPaths.cpp:
(WebCore::CSSParserFastPaths::maybeParseValue): Use isExposed.

* Source/WebCore/css/parser/CSSParserImpl.cpp:
(WebCore::CSSParserImpl::consumeDeclaration): Use isExposed.

* Source/WebCore/css/parser/CSSPropertyParser.cpp: Removed include of
HashTools.h.
(WebCore::cssPropertyID): Removed unneeded extra characters in buffer. Removed
code that adds a trailing NUL character. Use findCSSProperty.
(WebCore::isAppleLegacyCSSValueKeyword): Removed unnecessarily complex handling
of "-apple-wireless" prefix, simplifying making it match the other cases. Also
tweaked the logic to do less excess checking.
(WebCore::cssValueKeywordID): Removed code that adds a trailing NUL character.
SImplified the isAppleLegacyCSSValueKeyword section a little and removed a
misplaced comment. Use findCSSValueKeyword.
(WebCore::CSSPropertyParser::addProperty): Use isExposed.
(WebCore::consumeWillChange): Use isExposed.
(WebCore::CSSPropertyParser::parseSingleValue): Use isExposed.
(WebCore::CSSPropertyParser::parseCounterStyleDescriptor): Use isExposed.
(WebCore::CSSPropertyParser::parseFontFaceDescriptor): Use isExposed.
(WebCore::CSSPropertyParser::parseFontPaletteValuesDescriptor): Use isExposed.
(WebCore::CSSPropertyParser::consumeContainIntrinsicSizeShorthand): Use isExposed.

* Source/WebCore/css/typedom/ComputedStylePropertyMapReadOnly.cpp:
(WebCore::ComputedStylePropertyMapReadOnly::get const): Use isExposed.
(WebCore::ComputedStylePropertyMapReadOnly::getAll const): Use isExposed.
(WebCore::ComputedStylePropertyMapReadOnly::entries const): Use nameString.

* Source/WebCore/css/typedom/MainThreadStylePropertyMapReadOnly.cpp:
(WebCore::MainThreadStylePropertyMapReadOnly::get const): Use isExposed.
(WebCore::MainThreadStylePropertyMapReadOnly::getAll const): Use isExposed.

* Source/WebCore/css/typedom/StylePropertyMap.cpp:
(WebCore::StylePropertyMap::remove): Use isExposed.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::exposedComputedCSSPropertyIDs): Tweaked code since
computedPropertyIDs is now a std::array. Also decreased copying a little
using std::optional::emplace. We could do even better if we made a FixedVector
constructor that took a pair of iterators or a Span. Use isExposed.
(WebCore::Document::detachFromFrame): Fixed comment typo.

* Source/WebCore/html/CustomPaintImage.cpp: Use nameString.

* Source/WebCore/inspector/InspectorStyleSheet.cpp:
(WebCore::InspectorStyle::collectProperties const): Use allCSSProperties and isExposed.
Fixes a bug where we would not iterate over the last CSS property. Use nameString.
(WebCore::InspectorStyle::styleWithProperties const): Use isExposed and nameString.

* Source/WebCore/inspector/agents/InspectorAnimationAgent.cpp:
(WebCore::buildObjectForKeyframes): Use nameString.

* Source/WebCore/inspector/agents/InspectorCSSAgent.cpp:
(WebCore::InspectorCSSAgent::getSupportedCSSProperties): Use allCSSProperties,
isExposed, nameString, and allCSSValueKeywords.

* Source/WebCore/page/CaptionUserPreferencesMediaAF.cpp:
(WebCore::CaptionUserPreferencesMediaAF::captionsWindowCSS const): Use nameLiteral.
(WebCore::appendCSS): Turned this into a template to cut down on temporary strings.
(WebCore::CaptionUserPreferencesMediaAF::windowRoundedCornerRadiusCSS const):
Reorder arguments and stop using nameString when calling appendCSS.
(WebCore::CaptionUserPreferencesMediaAF::colorPropertyCSS const): Ditto.
(WebCore::CaptionUserPreferencesMediaAF::captionsTextEdgeCSS const): Ditto.
Also got rid of NeverDestroyed<const String> because we are appending these strings,
so we can just use literals.

* Source/WebCore/platform/HashTools.h: Remove Property, Value, findProperty, and findValue.

* Source/WebCore/platform/animation/Animation.cpp:
(WebCore::operator<<): Use nameLiteral.
* Source/WebCore/rendering/style/RenderStyleConstants.cpp:
(WebCore::operator<<): Ditto.

* Source/WebCore/style/PropertyCascade.cpp:
(WebCore::Style::PropertyCascade::lastDeferredPropertyResolvingRelated const):
Use relatedProperty.

* Source/WebCore/style/StyleBuilderCustom.h:
(WebCore::Style::BuilderCustom::applyValueWillChange): Use isExposed.

Canonical link: https://commits.webkit.org/256223@main
